central/auth/service/service_impl.go

```diff
@@ -89,3 +120,107 @@ func authStatusForID(id authn.Identity) (*v1.AuthStatus, error) {
+	config := request.GetConfig()
+	if err := s.validateAuthMachineToMachineConfig(ctx, config, true); err != nil {
+		return nil, err
+	}
+	config.Id = uuid.NewV4().String()
+	storageConfig, err := s.authDataStore.AddAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config))
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.AddAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(storageConfig)}, nil
+}
+
+func (s *serviceImpl) UpdateAuthMachineToMachineConfig(ctx context.Context, request *v1.UpdateAuthMachineToMachineConfigRequest) (*v1.UpdateAuthMachineToMachineConfigResponse, error) {
+	config := request.GetConfig()
+	if err := s.validateAuthMachineToMachineConfig(ctx, config, false); err != nil {
+		return nil, err
+	}
+
+	storageConfig, err := s.authDataStore.UpdateAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config))
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.UpdateAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(storageConfig)}, nil
+}
+
+func (s *serviceImpl) DeleteAuthMachineToMachineConfig(ctx context.Context, id *v1.ResourceByID) (*v1.Empty, error) {
+	if err := s.authDataStore.RemoveAuthM2MConfig(ctx, id.GetId()); err != nil {
+		return nil, errox.InvalidArgs.
+			Newf("could not delete auth machine to machine config with id %q", id.GetId()).CausedBy(err)
+	}
+	return &v1.Empty{}, nil
+}
+
+func (s *serviceImpl) ExchangeAuthMachineToMachineToken(_ context.Context, _ *v1.ExchangeAuthMachineToMachineTokenRequest) (*v1.ExchangeAuthMachineToMachineTokenResponse, error) {
+	return nil, errox.InvariantViolation.New("not yet implemented")
+}
+
+func (s *serviceImpl) validateAuthMachineToMachineConfig(ctx context.Context, config *v1.AuthMachineToMachineConfig,
+	skipIDCheck bool) error {
+	if config == nil {
+		return errox.InvalidArgs.New("empty config given")
+	}
+	if config.GetId() == "" && !skipIDCheck {
+		return errox.InvalidArgs.New("empty ID given")
+	}
+
+	duration, err := time.ParseDuration(config.GetTokenExpirationDuration())
+	if err != nil {
+		return errox.InvalidArgs.New("invalid token expiration duration given").CausedBy(err)
+	}
+
+	if duration < time.Minute || duration > 24*time.Hour {
+		return errox.InvalidArgs.Newf("token expiration must be between 1 minute and 24 hours, but was %s",
+			duration.String())
+	}
+
+	if config.GetType() == v1.AuthMachineToMachineConfig_GENERIC && config.GetGenericIssuerConfig().GetIssuer() == "" {
+		return errox.InvalidArgs.Newf("type %s was used, but no configuration for the issuer was given",
+			storage.AuthMachineToMachineConfig_GENERIC)
+	}
+
+	var roleValidationErrs error
+	for _, mapping := range config.GetMappings() {
+		// The user's context requires permissions to Access, which the role datastore requires as well. Hence,
+		// no need to elevate the context here.
+		_, exists, err := s.roleDataStore.GetRole(ctx, mapping.GetRole())
```

Comment:
I'd still prefer to do the error check explicitly in the validation logic for better UX and more clear error messages.

The safety net would then only come into play when during the time we validated the config a role will be removed from the store, which ultimately would lead to a error during insert of the config.

The logic itself and it's maintenance isn't a dealbreaker for me to justify removing it.

Reply:
I don't know the volume of these things we expect over time, but if we are going to use the foreign key then this seems like a wasted trip to the database.  The database is going to do this exact validation in a optimized way regardless and in race case could still fail.  IMO from a database traffic and performance perspective, I would prefer to not do this validation as long as the foreign key is on the column.

Reply:
I don't know the volume of these things we expect over time, but if we are going to use the foreign key then this seems like a wasted trip to the database.  The database is going to do this exact validation in a optimized way regardless and in race case could still fail.  IMO from a database traffic and performance perspective, I would prefer to not do this validation as long as the foreign key is on the column.

Reply:
This code looks gone so I'm resolving the thread.